src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/CommandBarViewModel.cs

```diff
@@ -0,0 +1,129 @@
+using CommunityToolkit.Mvvm.Input;
+using CommunityToolkit.Mvvm.Messaging;
+using Microsoft.CmdPal.UI.ViewModels.Messages;
+
+namespace Microsoft.CmdPal.UI.ViewModels;
+
+public partial class CommandBarViewModel : ObservableObject,
+    IRecipient<UpdateCommandBarMessage>
+{
+    public ICommandBarContext? SelectedItem
+    {
+        get => field;
+        set
+        {
+            field = value;
+            SetSelectedItem(value);
+        }
+    }
+
+    [ObservableProperty]
+    [NotifyPropertyChangedFor(nameof(HasPrimaryCommand))]
+    public partial CommandItemViewModel? PrimaryCommand { get; set; }
+
+    public bool HasPrimaryCommand => PrimaryCommand != null && PrimaryCommand.ShouldBeVisible;
+
+    [ObservableProperty]
+    [NotifyPropertyChangedFor(nameof(HasSecondaryCommand))]
+    public partial CommandItemViewModel? SecondaryCommand { get; set; }
+
+    public bool HasSecondaryCommand => SecondaryCommand != null;
+
+    [ObservableProperty]
+    public partial bool ShouldShowContextMenu { get; set; } = false;
+
+    [ObservableProperty]
+    public partial PageViewModel? CurrentPage { get; set; }
+
+    [ObservableProperty]
+    public partial ObservableCollection<CommandContextItemViewModel> ContextCommands { get; set; } = [];
+
+    public CommandBarViewModel()
+    {
+        WeakReferenceMessenger.Default.Register<UpdateCommandBarMessage>(this);
+    }
+
+    public void Receive(UpdateCommandBarMessage message) => SelectedItem = message.ViewModel;
+
+    private void SetSelectedItem(ICommandBarContext? value)
+    {
+        if (value != null)
+        {
+            PrimaryCommand = value.PrimaryCommand;
+            value.PropertyChanged += SelectedItemPropertyChanged;
+        }
+        else
+        {
+            if (SelectedItem != null)
+            {
+                SelectedItem.PropertyChanged -= SelectedItemPropertyChanged;
+            }
+
+            PrimaryCommand = null;
+        }
+
+        UpdateContextItems();
+    }
+
+    private void SelectedItemPropertyChanged(object? sender, System.ComponentModel.PropertyChangedEventArgs e)
+    {
+        switch (e.PropertyName)
+        {
+            case nameof(SelectedItem.HasMoreCommands):
+                UpdateContextItems();
+                break;
+        }
+    }
+
+    private void UpdateContextItems()
+    {
+        if (SelectedItem == null)
+        {
+            SecondaryCommand = null;
+            ShouldShowContextMenu = false;
+            return;
+        }
+
+        SecondaryCommand = SelectedItem.SecondaryCommand;
+
+        if (SelectedItem.MoreCommands.Count() > 1)
+        {
+            ShouldShowContextMenu = true;
+            ContextCommands = [.. SelectedItem.AllCommands];
+        }
+        else
+        {
+            ShouldShowContextMenu = false;
+        }
+    }
+
+    // InvokeItemCommand is what this will be in Xaml due to source generator
+    // this comes in when an item in the list is tapped
```

Comment:
Worth making these nice intellisense comments? Also not sure we need line above explaining how RelayCommands work

Reply:
genuinely I still don't think I totally understand relay commands, so it was helpful to me in debugging 😅